### PR TITLE
Support activity level in Yunmai scales

### PR DESCRIPTION
**Pull request by d3m3vilurr**
this patch needs to confirm from newer version users and fix the muscle value of
normal body users

a getMuscle function doesn't be related `sex`, only related `body_type`
in the original app, if body_type is 1(fitness body type), it should multiply
with 0.7.
if not (normal body type), it should multiply with 0.67

### CODE REVIEW

**Comment by d3m3vilurr** [comment]
this patch is fully implement version of the #521 
but it doesn't test in higher version scale, I don't know it's stable or not.

at least, I checked the original codeset that BLE packet reuse `body type` value; it should be 0(normal), 1(fitness), -1(unknown)

**Comment by d3m3vilurr** [comment]
an example of higher version device is #490 
an issue has packet sample; `0D 1F 14 02 00 5D 67 C1 4E 00 00 9E 15 1C BD 01 A1 08 C0 FE` (device version is `0x1F`)
IMO, we should do cross checking that `0x08C0` is same as `fat` in original app. :)

**Comment by oliexdev** [comment]
I write down some values from my last experiments with the original app in issue https://github.com/oliexdev/openScale/issues/71#issuecomment-380172378
At least for these values the last two bytes are the body fat for the yunmai mini devices...
Do we get the same values?

**Comment by d3m3vilurr** [comment]
<details><summary>checking scripts</summary>

```
def get_bmi(weight, height):
    return weight / (((height * height) / 100.0) / 100.0)

def get_fat(height, sex, age, weight, resistance, dev_ver=0x13, body_type=0):
    r = (resistance - 100) / 100.0
    h = height / 100.0
    if r >= 1:
        r = r ** 0.5
    if dev_ver >= 0x1F and body_type:
        bmi = get_bmi(weight, height)
        #print(bmi)
        b = bmi - 22.5
        a = abs(b)
        p = pow(r, a)
        imp = 0
        if sex == 1:
            i = 12.0 - r
        else:
            i = 22.8 - r
        fat = i + (p * (b / a))
    else:
        fat = weight * 1.5 / h / h + age * 0.08
        if sex == 1:
            fat -= 10.8
        fat = fat - 7.4 + r
    if 5.0 <= fat <= 75.0:
        return fat
    return 0.0

def get_water(fat):
    return ((100.0 - fat) * 0.726 * 100.0 + 0.5) / 100.0

def get_muscle(fat, body_type=0):
    muscle = (100.0 - fat) * 0.67
    if body_type:
        muscle = (100.0 - fat) * 0.7
    muscle = ((muscle * 100.0) + 0.5) / 100.0
    return muscle

def get_skeletal_muscle(fat, body_type=0):
    muscle = (100.0 - fat) * 0.53
    if body_type:
        muscle = (100.0 - fat) * 0.6
    muscle = ((muscle * 100.0) + 0.5) / 100.0;
    return muscle;

def get_bone_mass(height, sex, muscle, weight):
    h = height - 170.0
    if sex == 1:
        bone_mass = ((weight * (muscle / 100.0) * 4.0) / 7.0 * 0.22 * 0.6) + (h / 100.0)
    else:
        bone_mass = ((weight * (muscle / 100.0) * 4.0) / 7.0 * 0.22 * 0.6) + (h / 100.0)
    bone_mass = ((bone_mass * 10.0) + 0.5) / 10.0
    return bone_mass

def get_lean_body_mass(weight, fat):
    return weight * (100.0 - fat) / 100.0

def get_visceral_fat(sex, fat, age, body_type=0):
    f = fat
    a = age if 18 <= age <= 120 else 18
    if not body_type:
        if sex == 1:
            if a < 40:
                f -= 21.0
            elif a < 60:
                f -= 22.0
            else:
                f -= 24.0
        else:
            if a < 40:
                f -= 34.0
            elif a < 60:
                f -= 35.0
            else:
                f -= 36.0
        d = 1.4 if sex == 1 else 1.8
        if f > 0:
            d = 1.1
        f = (f / d) + 9.5
        return min(max(1.0, f), 30.0)
    else:
        if fat > 15.0:
            f = (fat - 15.0) / 1.1 + 12.0
        else:
            f = -1 * (15.0 - fat) / 1.4 + 12.0
        return min(max(1.0, f), 9.0)

def print_body_info(packet):
    packet = packet.replace(':', '')
    weight = int(packet[13*2:15*2], 16) / 100.0
    resistance = int(packet[15*2:17*2], 16)
    fat = int(packet[17*2:19*2], 16) / 100.0
    normal_body_fat = get_fat(175, 1, 25, weight, resistance)
    fit_body_fat = get_fat(175, 1, 25, weight, resistance, dev_ver=0x1F, body_type=1)
    normal_body_muscle, fit_body_muscle = get_muscle(fat), get_muscle(fat, body_type=1)
    normal_skel_muscle, fit_skel_muscle = get_skeletal_muscle(fat), get_skeletal_muscle(fat, body_type=1)
    water = get_water(fat)
    normal_body_bone, fit_body_bone = get_bone_mass(175, 1, normal_body_muscle, weight), get_bone_mass(175, 1, fit_body_muscle, weight)
    lbm = get_lean_body_mass(weight, fat)
    normal_body_vis, fit_body_vis = get_visceral_fat(1, fat, 25), get_visceral_fat(1, fat, 25, body_type=1)
    print('Weight: {:.2f}'.format(weight))
    print('Fat: {:.2f}'.format(fat))
    print('Expect fat: {:.2f} / {:.2f}'.format(normal_body_fat, fit_body_fat))
    print('Muscle: {:.2f} / {:.2f}'.format(normal_body_muscle, fit_body_muscle))
    print('Skeletal muscle: {:.2f} / {:.2f}'.format(normal_skel_muscle, fit_skel_muscle))
    print('Water: {:.2f}'.format(water))
    print('Protein: TODO')
    print('Bone mass: {:.2f} / {:.2f}'.format(normal_body_bone, fit_body_bone))
    print('Lean body mass: {:.2f}'.format(lbm))
    print('Visceral fat: {:.2f} / {:.2f}'.format(normal_body_vis, fit_body_vis))
```

</details>

first, I couldn't use `get_fat` function for the 0x1F packets in my test.
a script run with packet's fat value.
and IMO, packets aren't fitness body type.

Case1
```
0d:1f:14:02:00:59:c4:ff:cf:23:d0:b6:32:26:aa:00:9f:08:de:16  // 99,0 kg
22.9.2017 14:19  99,0 kg
Body fat 22,7
Muscle 51,8
Water 56,1
Protein 16,2
Bone mass 3,9
Resistance 0x9f (maybe!?)

>>> print_body_info('0d:1f:14:02:00:59:c4:ff:cf:23:d0:b6:32:26:aa:00:9f:08:de:16')
Weight: 98.98
Fat: 22.70
Expect fat: 32.87 / 11.42
Muscle: 51.80 / 54.11
Skeletal muscle: 40.97 / 46.38
Water: 56.12
Protein: TODO
Bone mass: 3.97 / 4.14
Lean body mass: 76.51
Visceral fat: 11.05 / 9.00
```

Case2
```
0d:1f:14:02:00:59:c5:3b:d8:23:d0:b6:32:28:ae:00:c2:09:c4:88  // 104,1 kg
22.9.2017 18:35 104,1 kg 
Body fat 25,0
Muscle 50,3% 
Water 54,5% 
Protein 16,1
bone mass 3,8
Resistance 0xc2 (maybe!?)

>>> print_body_info('0d:1f:14:02:00:59:c5:3b:d8:23:d0:b6:32:28:ae:00:c2:09:c4:88')
Weight: 104.14
Fat: 25.00
Expect fat: 35.75 / 11.55
Muscle: 50.26 / 52.51
Skeletal muscle: 39.76 / 45.01
Water: 54.45
Protein: TODO
Bone mass: 4.05 / 4.22
Lean body mass: 78.11
Visceral fat: 13.14 / 9.00
```

Case3
```
0d:1f:14:02:00:59:c5:f7:d6:23:d0:b6:32:26:3b:00:dd:08:f2:f9  // 97,9 kg
23.9.2017 07:36 97,9 kg 
Body fat 22,9
Muscle 51,7% 
Water 56,0% 
Protein 16,2
bone mass 4,0
Resistance 0xdd (maybe!?)

>>> print_body_info('0d:1f:14:02:00:59:c5:f7:d6:23:d0:b6:32:26:3b:00:dd:08:f2:f9')
Weight: 97.87
Fat: 22.90
Expect fat: 32.84 / 13.36
Muscle: 51.66 / 53.97
Skeletal muscle: 40.87 / 46.27
Water: 55.98
Protein: TODO
Bone mass: 3.91 / 4.08
Lean body mass: 75.46
Visceral fat: 11.23 / 9.00
```

Case4
```
0d:1f:14:02:00:59:c6:15:2f:23:d0:b6:32:26:95:00:a5:08:f2:37  // 98,8 kg
23.9.2017 10:02 98,8 kg 
Body fat 22,9
Muscle 51,7% 
Water 56,0% 
Protein 16,2
bone mass 3,9
Resistance 0xa5 (maybe!?)

>>> print_body_info('0d:1f:14:02:00:59:c6:15:2f:23:d0:b6:32:26:95:00:a5:08:f2:37')
Weight: 98.77
Fat: 22.90
Expect fat: 32.83 / 11.36
Muscle: 51.66 / 53.97
Skeletal uscle: 40.87 / 46.27
Water: 55.98
Protein: TODO
Bone mass: 3.95 / 4.12
Lean body mass: 76.15
Visceral fat: 11.23 / 9.00
```

**Comment by d3m3vilurr** [comment]
devices that version is less than `device_version >= 0x1E` really don't use `getFat` function.
I'm not sure that this `fitness body` value really exists and supports or not :p

BLE data reading part
```
   93     public static C4131w m17254a(String str, String str2, String str3) {
   94         C4131w wVar = new C4131w();
   95         try {
   96             int parseInt = Integer.parseInt(str.substring(2, 4), 16);
   97             int parseInt2 = Integer.parseInt(str.substring(8, 10));
   98             Date a = C3567ad.m17081a(Integer.valueOf(Integer.parseInt(str.substring(10, 18), 16)));
   99             int parseInt3 = Integer.parseInt(str.substring(18, 26), 16);
  100             float a2 = C3565ab.m17032a(((float) Integer.parseInt(str.substring(26, 30), 16)) * 0.01f, 2);
  101             int parseInt4 = Integer.parseInt(str.substring(30, 34), 16);
  102             float f = 0.0f;
  103             if (parseInt >= 30) {
  104                 f = ((float) Integer.parseInt(str.substring(34, 38), 16)) * 0.01f;
  105             }
  106             wVar.mo28502a(parseInt3);
  107             wVar.mo28503a(C2460bk.m7764e(StringUtils.defaultIfEmpty(str3, "")));
  108             wVar.mo28508b(StringUtils.defaultIfEmpty(str2, ""));
  109             wVar.mo28511c(StringUtils.defaultIfEmpty(str3, ""));
  110             wVar.mo28514d("9ABFC1D3-EA97-1758-7FBC-" + C2460bk.m7764e(StringUtils.defaultIfEmpty(str3, "")));
  111             wVar.mo28504a(a);
  112             wVar.mo28501a(a2);
  113             wVar.mo28507b(parseInt4);
  114             wVar.mo28513d(parseInt2);
  115             wVar.mo28506b(f);
  116             wVar.mo28516e(parseInt);
  117         } catch (NumberFormatException e) {
  118             e.printStackTrace();
  119         }
  120         return wVar;
```

data calculation part
```
     48         if (l >= 30) {
     49             formFat = wVar.mo28522k();
     50         } else {
>>   51             formFat = FormulaJNI.formFat(userBase.mo19851w(), userBase.mo19806e(), userBase.mo19849u(), f, (float) wVar.mo28518g(), l, userBase.mo19771M(), Locale.getDefault().getLanguage());
     52         }
```

**Comment by oliexdev** [comment]
Your "Yunmai activityLevel" is only 1 and 0 which is equal to the gender. An activity level should have more values rather than two. So I assumed that this is the gender. What do you think?

**Comment by d3m3vilurr** [comment]
no, it's not a gender.
`sex` velue is different variable.

lib:
```
Java_com_yunmai_scale_logic_config_FormulaJNI_getValue
          (int *jni_env,undefined4 j_object,int height,undefined4 age,undefined4 sex,float weight,
          undefined4 resistance,float fat,undefined4 device_version,int body_type,undefined4 locale)
```
in java
```
float[] formValue = FormulaJNI.formValue(userBase.mo19851w(), userBase.mo19806e(), userBase.mo19849u(), f, (float) wVar.mo28518g(), formFat, l, userBase.mo19771M(), Locale.getDefault().getLanguage());
```
UserBase
```
...
    539     public short mo19849u() {
    540         return this.f7607an;
    541     }
...
    714     public short mo19771M() {
    715         return this.f7583aB;
    716     }  
...
    788     public String toString() {
    789         return "UserBase [id=" + this.f7576U + ", userId=" + this.f7577V + ", bbsId=" + this.f7578W + ", userName=" + this.f7579X + ", phoneNo=" + this.f7580Y + ", eMail=" + this.f7581Z + ", weiBoNo=" + this.f7594aa + ", weiXinNo=" + this.f7595ab + ", qqNo=" + this
        .f7596ac + ", password=" + this.f7597ad + ", status=" + this.f7598ae + ", registerType=" + this.f7599af + ", nickName=" + this.f7600ag + ", realName=" + this.f7601ah + ", avatarUrl=" + this.f7602ai + ", birthday=" + this.f7606am + ", sex=" + this.f7607an + ", unit=
        " + this.f7608ao + ", height=" + this.f7609ap + ", waistLine=" + this.f7611ar + ", bust=" + this.f7612as + ",   basisWeight=" + this.f7614au + ", firstFat=" + this.f7613at + ", fristWeight=" + this.f7615av + ",relevanceName=" + this.f7619az + ", age=" + this.f7591a
        J + ", puid=" + this.f7618ay + ", description=" + this.f7603aj + ", indexImgUrl=" + this.f7604ak + ", exitDevice=" + this.f7582aA + ", bodyType=" + this.f7583aB + ", premiumEmail=" + this.f7584aC + ", premiumStatus=" + this.f7585aD + ", premiumBillType=" + this.f75
        86aE + ", premiumOrderNo=" + this.f7588aG + ", premiumTypeDesc=" + this.f7587aF + ", premiumGoodsSkuId=" + this.f7589aH + "]";
    790     }
...
```

more detail reversing info:

this value would change in the setting screen(but I can't find it. so it might be related scale machine or country)
```
   17             <RelativeLayout android:id="@+id/setting_userfeature" android:visibility="gone" style="@style/activity_settting_account_item">
   18                 <TextView android:text="@string/userfeature_dialog_title" style="@style/activitySettingItemTitle"/>
   19                 <com.yunmai.scaleen.component.CustomerSwitchButton android:id="@+id/setting_userfeature_switch" android:layout_width="@dimen/px175" android:layout_height="@dimen/px105" android:layout_marginRight="@dimen/px60" android:layout_alignParentRight="true" an
      droid:layout_centerVertical="true"/>
   20             </RelativeLayout>
```
and `userfeature_dialog_title` means 
```
resources/res/values/strings.xml:    <string name="userfeature_dialog_title">Athlete mode</string>
```
so technically, it's fitness info and value range is -1, 0 and 1.
only 1 is fitness mode. idk -1's meaning.

and usually, an original application uses `0`.

File: android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothYunmaiSE_Mini.java [hunk 1]
```diff
@@ -60,12 +60,13 @@ protected boolean onNextStep(int stepNr) {
...
                 byte sex = selectedUser.getGender().isMale() ? (byte)0x01 : (byte)0x02;
                 byte display_unit = selectedUser.getScaleUnit() == Converters.WeightUnit.KG ? (byte) 0x01 : (byte) 0x02;
+                byte body_type = (byte) YunmaiLib.toYunmaiActivityLevel(selectedUser.getActivityLevel());
 
                 byte[] user_add_or_query = new byte[]{
                         (byte) 0x0d, (byte) 0x12, (byte) 0x10, (byte) 0x01, (byte) 0x00, (byte) 0x00,
                         userId[0], userId[1], (byte) selectedUser.getBodyHeight(), sex,
                         (byte) selectedUser.getAge(), (byte) 0x55, (byte) 0x5a, (byte) 0x00,
-                        (byte)0x00, display_unit, (byte) 0x03, (byte) 0x00};
+                        (byte)0x00, display_unit, body_type, (byte) 0x00};
```
Comment (oliexdev) [inline android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothYunmaiSE_Mini.java]: If body_type is only 0 or 1 then why I wrote 3 here? Do you really know that this byte is for the body type? Unfortunately I can't test it anymore.

Comment (d3m3vilurr) [reply android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothYunmaiSE_Mini.java]: I check this part in original app. and yes, this byte is body type.
imo you've been copied this info from [gist](https://gist.github.com/conoro/f0c1d96c450a8f5cce70e2846c3686c4)
and.. I don't know how conoro and pwnall get the 0x03 value

Comment (oliexdev) [reply android_app/app/src/main/java/com/health/openscale/core/bluetooth/BluetoothYunmaiSE_Mini.java]: ok I see thanks

**Pull request merged**